yellowbox/extras/logstash.py

```diff
@@ -1,30 +1,300 @@
-from yellowbox.containers import get_ports, create_and_pull
-from yellowbox.subclasses import SingleContainerService, RunMixin
+__all__ = ['FakeLogstashService']
+_logger = logging.getLogger(__name__)
 
-__all__ = ['LogstashService', 'LOGSTASH_DEFAULT_PORT']
+_STOP_TIMEOUT = 5  # Timeout for stopping the service
 
-LOGSTASH_DEFAULT_PORT = 5959
+# Sent on internal connection to signal closing of background thread
+_CLOSE_SENTINEL = b"\0"
 
 
-class LogstashService(SingleContainerService, RunMixin):
-    def __init__(self, docker_client: DockerClient, image='logstash:7.8.1', **kwargs):
-        super().__init__(create_and_pull(
-            docker_client, image, publish_all_ports=True, detach=True,
-            ports={LOGSTASH_DEFAULT_PORT: None}
-        ), **kwargs)
+def _level_to_int(level: Union[str, int]) -> int:
+    """Convert a log level in str or int into an int.
 
-    def start(self, retry_spec: Optional[RetrySpec] = None):
-        # for now we just start the service, re-implementing this service is an ongoing ticket
-        return super().start(retry_spec)
+    Args:
+        level: Can be a string or int
 
-    def client_port(self):
-        return get_ports(self.container)[LOGSTASH_DEFAULT_PORT]
+    Returns:
+        level as int.
+    """
+    if not isinstance(level, int):
+        level = logging.getLevelName(level.upper())
 
-    def stop(self, signal='SIGKILL'):
-        # change in default
-        return super().stop(signal)
+    # In rare cases it might be false, but it shouldn't generally happen.
+    assert isinstance(level, int)
+    return level
+
+
+class FakeLogstashService(YellowService):
+    """Implements a fake logging service that closely resembles Logstash.
+
+    Accepts external TCP connections, with logs received in the "json_lines"
+    format.
+
+    Attributes:
+        records: A list of all records received. You are welcome to modify,
+        clear or iterate over this during runtime. New logs will be added in the
+        order they were received. Each record is a dict - for possible keys see
+        attribute docs.
+        port: Dynamic port external service should connect to. FakeLogstashService
+        automatically binds a free port during initialization unless chosen
+        otherwise.
+        local_host: Host to connect to from the local machine (=="localhost")
+        container_host: Host to connect to from inside containers.
+        encoding: Encoding of the json lines received. Defaults to utf-8 per
+        specification.
+        delimiter: Delimiter splitting between json objects. Defaults to b'\n'
+        per specification.
+
+    Example:
+        >>> import socket
+        >>> import time
+        >>> ls = FakeLogstashService()
+        >>> ls.start()
+        >>> s = socket.create_connection((ls.local_host, ls.port))  # Logstash Handler
+        >>> s.sendall(b'{"record": "value", "level": "ERROR"}\\n')
+        >>> time.sleep(0.01)  # Wait for service to process message
```

Comment:
The example should demonstrate a typical use case

```suggestion
        >>> ... # run a process that logs an error message with message "value"
```

Reply:
I'm using doctest for this as well...